presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTimeWithTimeZone.java

```diff
@@ -16,30 +16,64 @@
 import com.fasterxml.jackson.annotation.JsonValue;
 
 import java.time.Instant;
+import java.time.LocalTime;
 import java.time.ZoneId;
+import java.time.ZonedDateTime;
 import java.time.format.DateTimeFormatter;
 import java.util.Objects;
 import java.util.TimeZone;
 
 import static com.facebook.presto.spi.type.DateTimeEncoding.unpackMillisUtc;
 import static com.facebook.presto.spi.type.DateTimeEncoding.unpackZoneKey;
 
+/**
+ * This type represent instant of time in some day and timezone, where timezone
+ * is known and day is not.
+ */
 public final class SqlTimeWithTimeZone
 {
     private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm:ss.SSS VV");
 
+    /**
+     * As we dont have negative timestamps and max value is midnight in timezone that is ~(UTC -12:00),
+     * just not to care about precision we do use ~(UTC -24:00).
+     *
+     * It's all about protecting type from beeing used with full millisUTC.
+     */
+    private static final long MAX_MILLIS_TIME = 2 * 24 * 60 * 60 * 1000;
+
+    /**
+     * Milliseconds relative to midnight UTC of the given time (millisUtc < MAX_MILLIS_TIME).
+     */
     private final long millisUtc;
     private final TimeZoneKey timeZoneKey;
 
+    /**
+     * FIXME This is used only to determine last of TZ shifts for TIME
+     * This is workaround to avoid chaning semantic of millisUTC being part of SPI
+     * This is best information about current TZ we can get at the point of query
+     *
+     * eg. Asia/Katmandu was +5:30 at some point in a history and then was changed
+     *     to +5:45, without usage of this timestamp, we would use TZ shift valid
+     *     on 1970-01-01, which is invalid today.
+     */
+    private final long timezoneOffsetReferenceTimestampUtc = System.currentTimeMillis();
```

Comment:
I really was only thinking about making this more static (like having singleton for all Presto that will just provide that value). Just to avoid potential change in some weird scenarios.

Think about that, if we would call System.currentTimeMillis in toString, we would have DTO with toString (potentially) morphing over time.

I'm strongly against making this calculation later and more "dynamic", I would rather talk about making it earlier and more "static".